pkg/net/ip.go

```diff
@@ -28,76 +27,87 @@ const (
 	IPv6len = 16
 )
 
-func netGetIP(ip cue.Value) (goip net.IP) {
+func netGetIP(ip cue.Value) (goip netip.Addr) {
 	switch ip.Kind() {
 	case cue.StringKind:
 		s, err := ip.String()
 		if err != nil {
-			return nil
+			return netip.Addr{}
 		}
-		goip := net.ParseIP(s)
-		if goip == nil {
-			return nil
+		goip, err := netip.ParseAddr(s)
+		if err != nil {
+			return netip.Addr{}
 		}
 		return goip
 
 	case cue.BytesKind:
 		b, err := ip.Bytes()
 		if err != nil {
-			return nil
+			return netip.Addr{}
 		}
-		goip := net.ParseIP(string(b))
-		if goip == nil {
-			return nil
+		goip, err := netip.ParseAddr(string(b))
+		if err != nil {
+			return netip.Addr{}
 		}
 		return goip
 
 	case cue.ListKind:
 		iter, err := ip.List()
 		if err != nil {
-			return nil
+			return netip.Addr{}
 		}
+		var bytes []byte
 		for iter.Next() {
 			v, err := iter.Value().Int64()
 			if err != nil {
-				return nil
+				return netip.Addr{}
 			}
 			if v < 0 || 255 < v {
-				return nil
+				return netip.Addr{}
 			}
-			goip = append(goip, byte(v))
+			bytes = append(bytes, byte(v))
+		}
+		goip, ok := netip.AddrFromSlice(bytes)
+		if !ok {
+			return netip.Addr{}
 		}
 		return goip
 
 	default:
 		// TODO: return canonical invalid type.
-		return nil
+		return netip.Addr{}
 	}
 }
 
-func netGetIPCIDR(ip cue.Value) (gonet *net.IPNet, err error) {
+func netGetIPCIDR(ip cue.Value) (gonet *netip.Prefix, err error) {
 	switch ip.Kind() {
 	case cue.StringKind:
 		s, err := ip.String()
 		if err != nil {
 			return nil, err
 		}
-		_, gonet, err := net.ParseCIDR(s)
+		cidr, err := netip.ParsePrefix(s)
 		if err != nil {
 			return nil, err
 		}
-		return gonet, nil
+		if !cidr.IsValid() {
+			return nil, err
```

Comment:
this would result in `nil, nil`, right?

I suspect this shouldn't be possible; it would seem odd for ParsePrefix to return an invalid prefix with a nil error.

If it is possible, then I suspect what we want to do is return `nil, fmt.Errorf(...)`.

Reply:
Good catch. `netip.ParsePrefix` returns an error rather than an invalid `Prefix` so that check is redundant. I'll remove it.

Reply:
Fixed.